translated/tech/20181230-building-immutable-data-structures-in-go.md

```diff
@@ -0,0 +1,304 @@
+    return p.WithFavoriteColorAt(0, "red")
+}
+```
+
+想要了解更多切片的妙用参见这篇牛逼的wiki：https://github.com/golang/go/wiki/SliceTricks
+
+## 构造函数
+
+某些情况下，我们会假设结构体的默认值是合理的。但是，强烈建议总是创建构造函数，一旦将来需要改变默认值时，我们只需要改动一个地方：
+
+```go
+func NewPerson() Person {
+    return Person{}
+}
+```
+
+你可以随心所欲地实例化 `Person`，但个人偏爱总是通过 setter 来执行状态变换从而保持代码一致性：
+
+```go
+func NewPerson() Person {
+    return Person{}.
+        WithName("No Name")
+}
+```
+
+## 接口 (Interface)
+
+到现在为止，我们使用的还是公有的结构体。任由这些结构体方法摆布之下，加上创建 mock 可能会引发非预期的副作用，测试起来会很痛苦。
+
+我们可以创建一个同名的接口，并把相应的结构体重命名为 `person` 使之私有化：
+
+```go
+type Person interface {
+    WithName(name string) Person
+    Name() string
+    WithFavoriteColors(favoriteColors []string) Person
+    NumFavoriteColors() int
+    FavoriteColorAt(i int) string
+    WithFavoriteColorAt(i int, favoriteColor string) Person
+}
+
+type person struct {
+    name           string
+    favoriteColors []string
+}
+```
+
+我们现在就可以只重写想要替换的逻辑来创建测试 mock：
+
+```go
+type personMock struct {
+    Person
+    receivedNewColor string
+}
+
+func (m personMock) WithFavoriteColorAt(i int, favoriteColor string) Person {
+    m.receivedNewColor = favoriteColor
+    return m
+}
+```
+
+测试代码样例如下：
+
+```go
+mock := personMock{}
+result := updateFavoriteColors(mock)
+
+result.(personMock).receivedNewColor // "red"
+```
+
+## 记录变化
+
+如我早前所言，完整的状态转换非常有益于调试，而且我们可以 wither 来挂入钩子的方式捕捉到所有或部分变换过程：
+
+```go
+func (p person) nextState() Person {
+    fmt.Printf("nextState: %#+v\n", p)
+    return p
+}
+
+func (p person) WithName(name string) Person {
+    p.name = name
+    return p.nextState() // <- Use "nextState" whenever you return.
+}
+```
+
+对于更加复杂的逻辑或个人偏好，你也可以采用 `defer` 的方式：
+
+```go
+func (p person) WithFavoriteColors(favoriteColors []string) Person {
+    defer func() {
+        p.nextState()
+    }()
+
+    p.favoriteColors = favoriteColors
+    return p
+}
+```
+
+这样变换就可看到了：
+
+```bash
+nextState: main.person{name:"No Name", favoriteColors:[]string(nil)}
+nextState: main.person{name:"Elliot", favoriteColors:[]string(nil)}
+nextState: main.person{name:"Elliot", favoriteColors:[]string{"black", "blue"}}
+```
+
+你可以添加更多诸如此类的信息。例如，时间戳、栈追踪记录和其他自定义的上下文信息来使得调试更加容易。
+
+## 历史及回滚
+
+除了打印变化之外，我们还可以收集这些状态作为历史：
+
+```go
+type Person interface {
+    // ...
+    AtVersion(version int) Person
+}
+
+type person struct {
+    // ...
+    history        []person
+}
+
+func (p *person) nextState() Person {
+    p.history = append(p.history, *p)
+    return *p
+}
+
+func (p person) AtVersion(version int) Person {
+    return p.history[version]
+}
+
+func main() {
+    me := NewPerson().
+        WithName("Elliot").
+        WithFavoriteColors([]string{"black", "blue"})
+
+    // We discard the result, but it will be put into the history.
+    updateFavoriteColors(me)
+
+    fmt.Printf("%s\n", me.AtVersion(0).Name())
+    fmt.Printf("%s\n", me.AtVersion(1).Name())
+}
+
+// No Name
+// Elliot
+```
+
+这对最后需要进行的审查是非常有用的。记录所有日志打印的历史应对后续异常的场景也是非常有用的，或者让历史随实例消亡即可。
```

Comment:
It’s also useful to record all the history to be logged if something goes wrong later, otherwise the history can be just discarded with the instance.
记录所有日志打印的历史应对后续异常的场景也是非常有用的，或者让历史随实例消亡即可。
后面的 otherwise 这一句感觉与原文有出入
记录所有历史（状态）到日志系统对后续异常的场景也是非常有用的，否则历史（状态）仅随实例（消失）而被丢弃

Reply:
我喜欢前半句的译法，对于后半句的话，看看这样的解释是否合理哈

有需要的话，保存历史状态可以用来处理异常场景，**但是如果没有需要的话**，让历史随实例销毁即可

我的粗暴理解是这样的：对于需要打日志的话，可以利用 wither，如果不用的话，就不要在 wither 里面打日志，没有记录的话，实例一挂历史自然飘散~

Reply:
可以的我感觉

Reply:
这里的两句话都有"非常有用"，感觉不好，我调整成了

这相当利于最后进行审查。记录所有日志打印的历史对处理后续异常的场景也是非常有用的，如果不需要的话，让历史随实例消亡即可。

辛苦再帮忙审一波~